### PR TITLE
DAOS-11136 object: Check ec parity leader in key query

### PR DESCRIPTION
**Pull request by wangdi1**
Choose right parity shard for non-ec parity rotation case.

Only choose minimum shards for ec parity rotation key query task.

Required-githooks: true

Signed-off-by: Di Wang <di.wang@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -6474,15 +6474,16 @@ dc_obj_query_key(tse_task_t *api_task)
 	for (i = grp_idx; i < grp_idx + grp_nr; i++) {
 		int start_shard;
 		int j;
+		int shard_cnt = 0;
 
 		/* Try leader for current group */
 		if (!obj_is_ec(obj) || (obj_is_ec(obj) && !obj->cob_ec_parity_rotate)) {
```
Comment (Nasf-Fan) [inline src/object/cli_obj.c:6480]: Some redundant check for EC?
if (!obj_is_ec(obj) || !obj->cob_ec_parity_rotate)

File: src/object/cli_obj.c [hunk 2]
```diff
@@ -6506,14 +6507,26 @@ dc_obj_query_key(tse_task_t *api_task)
 
 non_leader:
 		/* Then Try non-leader shards */
+		D_ASSERT(obj_is_ec(obj));
 		start_shard = i * obj_get_grp_size(obj);
 		D_DEBUG(DB_IO, DF_OID" EC needs to try all shards for group %d.\n",
 			DP_OID(obj->cob_md.omd_id), i);
 		for (j = start_shard; j < start_shard + daos_oclass_grp_size(&obj->cob_oca); j++) {
+			if (obj_shard_is_invalid(obj, j, DAOS_OBJ_RPC_QUERY_KEY))
```
Comment (Nasf-Fan) [inline src/object/cli_obj.c:6515]: Does the shard in "reintegrating" is valid or not? obj_shard_is_invalid() seems not exclude such target?

Comment (wangdi1) [reply src/object/cli_obj.c:6515]: it is using oclass_grp_size here, so reintegrating shard is not within this range.

File: src/object/cli_obj.c [hunk 3]
```diff
@@ -6506,14 +6507,26 @@ dc_obj_query_key(tse_task_t *api_task)
 
 non_leader:
 		/* Then Try non-leader shards */
+		D_ASSERT(obj_is_ec(obj));
 		start_shard = i * obj_get_grp_size(obj);
```
Comment (liuxuezhao) [inline src/object/cli_obj.c:6511]: here the start shard seems possibly incorrect if parity shard rotate enabled? (as it possibly is not the first data shard)

Comment (liuxuezhao) [reply src/object/cli_obj.c:6511]: ok, seems fine here.

**Review by liuxuezhao — APPROVED** [review]

File: src/object/cli_obj.c (hunk 1 above)
Comment (Nasf-Fan) [reply src/object/cli_obj.c:6480]: "if (!obj->cob_ec_parity_rotate)" seems enough.

**Review by Nasf-Fan — APPROVED** [review]

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Conflicts with latest master so will need to resolve to land.

**Comment by wangdi1** [comment]
failure is due to DAOS-12031.

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**